app/helpers/lti_helper.rb

```diff
@@ -0,0 +1,115 @@
+      end
+    end
+    if user_data.empty?
+      raise 'No user data returned from lms'
+    end
+    user_data.each do |lms_user|
+      markus_user = EndUser.find_by(user_name: lms_user[:user_name])
+      if markus_user.nil? && can_create_users # allowed_to?(:manage?, current_user, with: Admin::UserPolicy)
+        markus_user = EndUser.create!(lms_user.except(:lti_user_id))
+      elsif markus_user.nil? && !can_create_users
+        errors.append("MarkUs user #{lms_user[:user_name]} not found")
+        next
+      end
+      course_role = Student.find_by(user: markus_user, course: course)
+      if course_role.nil? && can_create_roles # allowed_to?(:manage?, role, with: RolePolicy)
+        course_role = Student.create!(user: markus_user, course: lti_deployment.course)
+      end
+      next if course_role.nil?
+      lti_user = LtiUser.find_or_initialize_by(user: markus_user, lti_client: lti_deployment.lti_client)
+      lti_user.update!(lti_user_id: lms_user[:lti_user_id])
+    end
+    errors
+  end
+
+  def grade_sync(lti_deployment, assessment)
+    scopes = [LtiDeployment::LTI_SCOPES[:score], LtiDeployment::LTI_SCOPES[:results]]
+    lti_deployment.create_or_update_lti_assessment(assessment)
+    auth_data = lti_deployment.lti_client.get_oauth_token(scopes)
+    line_item = lti_deployment.lti_line_items.find_by!(assessment: assessment)
+    results_uri = URI("#{line_item.lti_line_item_id}/results")
+    result_req = Net::HTTP::Get.new(results_uri)
+    curr_results = lti_deployment.send_lti_request!(result_req, results_uri, auth_data, scopes)
+    curr_results = JSON.parse(curr_results.body)
+    score_uri = URI("#{line_item.lti_line_item_id}/scores")
+    req = Net::HTTP::Post.new(score_uri)
+
+    if assessment.is_a?(Assignment)
+      marks = get_assignment_marks(lti_deployment, assessment)
+    else
+      marks = get_grade_entry_form_marks(lti_deployment, assessment)
+    end
+    marks.each do |lti_user_id, mark|
+      # Only send if the mark has not been previously sent to the LMS
+      # or if the mark differs from the LMS mark.
+      marked_by_lms = curr_results.find { |result| result['userId'] == lti_user_id }
+      if marked_by_lms.nil? || marked_by_lms['resultScore'] != mark
+        payload = {
+          timestamp: Time.current.iso8601,
+          scoreGiven: mark,
+          scoreMaximum: assessment.max_mark.to_f,
+          activityProgress: 'Completed',
+          gradingProgress: 'FullyGraded',
+          userId: lti_user_id
+        }
+        req.set_form_data(payload)
+        lti_deployment.send_lti_request!(req, score_uri, auth_data, scopes)
+      end
+    end
+  end
+
+  # Returns a hash mapping lti_user_id to marks
+  # for each released mark where the user has an lti_user_id
+  def get_assignment_marks(lti_deployment, assignment)
+    marks = assignment.released_marks
+    mark_data = {}
+    lti_users = LtiUser.where(lti_client: lti_deployment.lti_client)
+    marks.each do |mark|
```

Comment:
I feel like this could be done in a single database query....

Reply:
Or at least make sure of the Result.total_marks method with a list of result ids that you can get in a single query